### DISCUSSION TITLE
[d3] 7.4.0 doesn't pin versions of its `@types/d3-*` dependencies

### DISCUSSION THREAD

**Discussion by dimfeld** [DefinitelyTyped/DefinitelyTyped#62436] [Issues with a @types package] [answered]
`@types/d3` 7.x uses the `"*"` version specifier for all of its dependencies (See https://unpkg.com/@types/d3@7.4.0/package.json), which means that package managers fail to upgrade the various dependent packages (e.g. `@types/d3-array`) when upgrading `@types/d3`. 

In my particular case, we upgraded from 5.x to 7.4.0. But, for example, `@types/d3-array` remained at the old version 1.2.9 instead of changing to the new version 3.0.3.

This appears to be because all the `paths` entries were removed from the `tsconfig.json` in https://github.com/DefinitelyTyped/DefinitelyTyped/commit/e6db7d50b6e1e96ccf07352cf5677c18c02d87fc, and so the explicit version pinning went away.

Do we just need to add back the `paths` in `tsconfig.json`? It wasn't clear to me from the README if you can reference the "latest" version in the paths though, so I'm not sure if this is actually the right thing to do or if it's more complex.

I've just worked around this in my project by forcing all the `@types/d3-*` to the appropriate versions, so this isn't actually affecting me for now, but I'm hoping there's a solution to make future upgrades easier. Thanks!

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment, marked as answer]
I'm not aware of a good solution here, I believe this is just an unfortunate reality of how DefinitelyTyped works.

**Reply by dimfeld** [reply]
Ok, thanks for the quick response!

**Reply by Semigradsky** [reply]
@Methuselah96 Why we can't change `"@types/d3-array": "*"` to `"@types/d3-array": "3"` and so on?

**Reply by Methuselah96** [reply]
The `package.json` and the version ranges that are used in the `package.json` are generated by the DefinitelyTyped tooling. I don't know of a way to override the generated version ranges offhand, but that doesn't mean it's not possible. Maybe you could try opening a PR where `@types/d3` has its own `package.json` that manually specifies the version ranges to see if it's possible to do so?

**Reply by Methuselah96** [reply]
Also to clarify my answer to the original question, the path entries went away because they were now referencing the latest version and not an older version. I'm not sure that it's possible to declare the `paths` so it points to the latest version and I'm not sure it would have the desired effect of avoiding `*`, but feel free to experiment with it to be sure.
